pkg/registry/core/namespace/storage/storage.go

```diff
@@ -185,18 +185,27 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
 					existingNamespace.Status.Phase = api.NamespaceTerminating
 				}
 
-				// Remove orphan finalizer if options.OrphanDependents = false.
-				if options.OrphanDependents != nil && *options.OrphanDependents == false {
-					// remove Orphan finalizer.
-					newFinalizers := []string{}
-					for i := range existingNamespace.ObjectMeta.Finalizers {
-						finalizer := existingNamespace.ObjectMeta.Finalizers[i]
-						if string(finalizer) != metav1.FinalizerOrphanDependents {
-							newFinalizers = append(newFinalizers, finalizer)
-						}
+				// remove orphan and foregroundDeletion first, add them back if needed
+				newFinalizers := []string{}
+				for _, f := range existingNamespace.Finalizers {
+					if f == metav1.FinalizerOrphanDependents || f == metav1.FinalizerDeleteDependents {
+						continue
 					}
-					existingNamespace.ObjectMeta.Finalizers = newFinalizers
+					newFinalizers = append(newFinalizers, f)
 				}
+
+				// add orphan finalizer if we need to orphan this namespace.
+				if (options.OrphanDependents != nil && *options.OrphanDependents) ||
+					(options.PropagationPolicy != nil && *options.PropagationPolicy == metav1.DeletePropagationOrphan) {
+					newFinalizers = append(newFinalizers, metav1.FinalizerOrphanDependents)
+				}
+
+				// add foregroundDeletion if the options says foreground deletion.
+				if options.PropagationPolicy != nil && *options.PropagationPolicy == metav1.DeletePropagationForeground {
+					newFinalizers = append(newFinalizers, metav1.FinalizerDeleteDependents)
+				}
+
+				existingNamespace.ObjectMeta.Finalizers = newFinalizers
```

Comment:
Check if finalizers have changed to avoid unnecessary update. https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L740-L744

Reply:
Also this is not the logic of the generic store, which will preserve the existing finalizers in the object is the deleteOptions doesn't explicitly set the mode of the deletion. See https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L619-L623 and https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L669-L673

Reply:
Also, could you double check how the namespace controller will react if the namespace deletion is blocked by others?

The old code doesn't honor the foregroundDeletion, which could cause the namespace deletion to be blocked.

Reply:
Ok, thanks for catching this. Now the behavior is changed to preserve finalizers if the delete options does not say explicitly.

> could you double check how the namespace controller will react if the namespace deletion is blocked by others?

Checking now.

Reply:
@caesarxuchao: Namespace controller does not care about finalizers in object meta too much, it only cares about the finalizer `kubernetes` in the spec.

If `kubernetes` finalizer is removed, and it gets an event, it just send a delete call again.